File: apps/api/app/api/v1/routes/trips.py
from __future__ import annotations

from dataclasses import asdict
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, HTTPException, Response, UploadFile, status
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.auth.admin import require_admin
from app.db.session import get_session
from app.models.admin_user import AdminUser
from app.models.trip import Trip
from app.models.trip_comment import TripComment
from app.models.trip_image import TripImage
from app.schemas.comment import CommentRead
from app.schemas.trip import TripCreate, TripRead, TripUpdate
from app.schemas.trip_image import TripImageOrderUpdate, TripImageRead
from app.services.gpx_polylines import build_trip_polyline_from_gpx
from app.services.image_uploads import (
    create_uploaded_image,
    delete_uploaded_image_files,
    rotate_uploaded_image,
)

router = APIRouter()


def _non_activity_image_payload(uploaded: object) -> dict[str, object]:
    payload = asdict(uploaded)
    payload.pop("capture_datetime_local", None)
    payload.pop("timezone", None)
    payload.pop("capture_datetime_utc", None)
    payload.pop("capture_timezone_source", None)
    payload.pop("capture_datetime_source", None)
    payload.pop("gps_datetime_utc", None)
    payload.pop("gps_timezone", None)
    return payload


def _to_trip_read(trip: Trip) -> TripRead:
    return TripRead.model_validate(trip)


async def _get_trip_or_404(session: AsyncSession, trip_id: int) -> Trip:
    stmt = (
        select(Trip)
        .options(selectinload(Trip.comments), selectinload(Trip.images))
        .where(Trip.id == trip_id)
    )
    trip = (await session.scalars(stmt)).first()
    if trip is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip not found")
    return trip


@router.get("", response_model=list[TripRead])
async def list_trips(
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[TripRead]:
    stmt = (
        select(Trip)
        .options(selectinload(Trip.comments), selectinload(Trip.images))
        .order_by(Trip.start_date.desc().nullslast(), Trip.created_at.desc())
    )
    trips = (await session.scalars(stmt)).all()
    return [_to_trip_read(trip) for trip in trips]


@router.post("", response_model=TripRead, status_code=status.HTTP_201_CREATED)
async def create_trip(
    payload: TripCreate,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TripRead:
    trip = Trip(**payload.model_dump())
    session.add(trip)
    await session.commit()
    trip = await _get_trip_or_404(session, trip.id)
    return _to_trip_read(trip)


@router.get("/{trip_id}", response_model=TripRead)
async def get_trip(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TripRead:
    trip = await _get_trip_or_404(session, trip_id)
    return _to_trip_read(trip)


@router.patch("/{trip_id}", response_model=TripRead)
async def update_trip(
    trip_id: int,
    payload: TripUpdate,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TripRead:
    trip = await _get_trip_or_404(session, trip_id)

    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(trip, field, value)

    await session.commit()
    trip = await _get_trip_or_404(session, trip_id)
    return _to_trip_read(trip)


@router.post("/{trip_id}/gpx", response_model=TripRead)
async def upload_trip_gpx(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
    file: Annotated[UploadFile, File(...)],
    compress: Annotated[bool, Form()] = False,
) -> TripRead:
    trip = await _get_trip_or_404(session, trip_id)

    payload = await file.read()
    if not payload:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Uploaded GPX file is empty"
        )

    trip.planned_path_polyline = build_trip_polyline_from_gpx(payload, compress=compress)
    await session.commit()
    trip = await _get_trip_or_404(session, trip_id)
    return _to_trip_read(trip)


@router.delete("/{trip_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_trip(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> Response:
    trip = await session.get(Trip, trip_id)
    if trip is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip not found")

    await session.delete(trip)
    await session.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)


def _delete_trip_image_files(image: TripImage) -> None:
    delete_uploaded_image_files(
        storage_path=image.storage_path,
        thumbnail_storage_path=image.thumbnail_storage_path,
        tiny_thumbnail_storage_path=image.tiny_thumbnail_storage_path,
    )


def _delete_trip_map_card_files(trip: Trip) -> None:
    if not trip.map_card_storage_path:
        return

    delete_uploaded_image_files(
        storage_path=trip.map_card_storage_path,
        thumbnail_storage_path="",
    )


@router.post(
    "/{trip_id}/images", response_model=list[TripImageRead], status_code=status.HTTP_201_CREATED
)
async def upload_trip_images(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
    files: Annotated[list[UploadFile], File(...)],
    resize_mode: Annotated[str, Form()] = "keep",
    resize_width: Annotated[int | None, Form()] = None,
    resize_height: Annotated[int | None, Form()] = None,
) -> list[TripImageRead]:
    await _get_trip_or_404(session, trip_id)

    if not files:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="No files were uploaded"
        )

    if resize_mode not in {"keep", "resize"}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resize mode")

    if resize_mode == "resize" and (
        resize_width is None or resize_height is None or resize_width <= 0 or resize_height <= 0
    ):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Resize width and height must be positive integers",
        )

    current_max_position = await session.scalar(
        select(func.max(TripImage.position)).where(TripImage.trip_id == trip_id)
    )
    next_position = (current_max_position or 0) + 1 if current_max_position is not None else 0

    created_images: list[TripImage] = []
    try:
        for file in files:
            uploaded = await create_uploaded_image(
                upload=file,
                resize_mode=resize_mode,
                resize_width=resize_width,
                resize_height=resize_height,
                storage_prefix=f"trips/{trip_id}",
            )
            image = TripImage(
                trip_id=trip_id,
                position=next_position,
                **_non_activity_image_payload(uploaded),
            )
            next_position += 1
            session.add(image)
            created_images.append(image)

        await session.commit()
    except Exception:
        await session.rollback()
        for image in created_images:
            _delete_trip_image_files(image)
        raise

    for image in created_images:
        await session.refresh(image)

    return [TripImageRead.model_validate(image) for image in created_images]


@router.post(
    "/{trip_id}/map-card-image", response_model=TripRead, status_code=status.HTTP_201_CREATED
)
async def upload_trip_map_card_image(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
    file: Annotated[UploadFile, File(...)],
    resize_mode: Annotated[str, Form()] = "keep",
    resize_width: Annotated[int | None, Form()] = None,
    resize_height: Annotated[int | None, Form()] = None,
) -> TripRead:
    trip = await _get_trip_or_404(session, trip_id)

    if resize_mode not in {"keep", "resize"}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid resize mode")

    if resize_mode == "resize" and (
        resize_width is None or resize_height is None or resize_width <= 0 or resize_height <= 0
    ):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Resize width and height must be positive integers",
        )

    previous_storage_path = trip.map_card_storage_path
    uploaded = await create_uploaded_image(
        upload=file,
        resize_mode=resize_mode,
        resize_width=resize_width,
        resize_height=resize_height,
        storage_prefix=f"trips/{trip_id}/map-card",
        create_thumbnails=False,
    )

    try:
        trip.map_card_storage_path = uploaded.storage_path
        trip.map_card_image_url = uploaded.image_url
        trip.map_card_width = uploaded.width
        trip.map_card_height = uploaded.height
        trip.map_card_content_type = uploaded.content_type
        await session.commit()
    except Exception:
        await session.rollback()
        delete_uploaded_image_files(
            storage_path=uploaded.storage_path,
            thumbnail_storage_path="",
        )
        raise

    if previous_storage_path:
        delete_uploaded_image_files(
            storage_path=previous_storage_path,
            thumbnail_storage_path="",
        )

    trip = await _get_trip_or_404(session, trip_id)
    return _to_trip_read(trip)


@router.delete("/{trip_id}/map-card-image", response_model=TripRead)
async def delete_trip_map_card_image(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TripRead:
    trip = await _get_trip_or_404(session, trip_id)
    _delete_trip_map_card_files(trip)
    trip.map_card_storage_path = None
    trip.map_card_image_url = None
    trip.map_card_width = None
    trip.map_card_height = None
    trip.map_card_content_type = None
    await session.commit()
    trip = await _get_trip_or_404(session, trip_id)
    return _to_trip_read(trip)


@router.patch("/{trip_id}/map-card-image/rotate", response_model=TripRead)
async def rotate_trip_map_card_image(
    trip_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TripRead:
    trip = await _get_trip_or_404(session, trip_id)
    if not trip.map_card_storage_path:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Trip map card image not found"
        )

    rotated = rotate_uploaded_image(
        storage_path=trip.map_card_storage_path,
        thumbnail_storage_path=None,
        tiny_thumbnail_storage_path=None,
        image_url=trip.map_card_image_url or "",
        thumbnail_url=None,
        tiny_thumbnail_url=None,
        content_type=trip.map_card_content_type or "image/jpeg",
        original_filename=None,
        gps_latitude=None,
        gps_longitude=None,
        capture_datetime_local=None,
        timezone=None,
        capture_datetime_utc=None,
        capture_timezone_source="unknown",
        capture_datetime_source="unknown",
        gps_datetime_utc=None,
        gps_timezone=None,
        create_thumbnails=False,
    )

    trip.map_card_storage_path = rotated.storage_path
    trip.map_card_image_url = rotated.image_url
    trip.map_card_width = rotated.width
    trip.map_card_height = rotated.height
    trip.map_card_content_type = rotated.content_type

    await session.commit()
    trip = await _get_trip_or_404(session, trip_id)
    return _to_trip_read(trip)


@router.delete("/{trip_id}/comments/{comment_id}", response_model=CommentRead)
async def delete_trip_comment(
    trip_id: int,
    comment_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> CommentRead:
    comment = await session.get(TripComment, comment_id)
    if comment is None or comment.trip_id != trip_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip comment not found")

    payload = CommentRead.model_validate(comment)
    await session.delete(comment)
    await session.commit()
    return payload


@router.delete("/{trip_id}/images/{image_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_trip_image(
    trip_id: int,
    image_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> Response:
    image = await session.get(TripImage, image_id)
    if image is None or image.trip_id != trip_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip image not found")

    _delete_trip_image_files(image)
    await session.delete(image)
    await session.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.put("/{trip_id}/images/order", response_model=list[TripImageRead])
async def reorder_trip_images(
    trip_id: int,
    payload: TripImageOrderUpdate,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[TripImageRead]:
    stmt = (
        select(TripImage)
        .where(TripImage.trip_id == trip_id)
        .order_by(TripImage.position.asc(), TripImage.created_at.asc(), TripImage.id.asc())
    )
    images = list((await session.scalars(stmt)).all())
    if not images:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip has no images")

    current_ids = [image.id for image in images]
    ordered_ids = payload.ordered_image_ids
    if len(ordered_ids) != len(current_ids) or set(ordered_ids) != set(current_ids):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Image order does not match trip images"
        )

    images_by_id = {image.id: image for image in images}
    for index, image_id in enumerate(ordered_ids):
        images_by_id[image_id].position = index

    await session.commit()
    ordered_images = [images_by_id[image_id] for image_id in ordered_ids]
    return [TripImageRead.model_validate(image) for image in ordered_images]


@router.patch("/{trip_id}/images/{image_id}/rotate", response_model=TripImageRead)
async def rotate_trip_image(
    trip_id: int,
    image_id: int,
    _: Annotated[AdminUser, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TripImageRead:
    image = await session.get(TripImage, image_id)
    if image is None or image.trip_id != trip_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip image not found")

    rotated = rotate_uploaded_image(
        storage_path=image.storage_path,
        thumbnail_storage_path=image.thumbnail_storage_path,
        tiny_thumbnail_storage_path=image.tiny_thumbnail_storage_path,
        image_url=image.image_url,
        thumbnail_url=image.thumbnail_url,
        tiny_thumbnail_url=image.tiny_thumbnail_url,
        content_type=image.content_type,
        original_filename=image.original_filename,
        gps_latitude=image.gps_latitude,
        gps_longitude=image.gps_longitude,
        capture_datetime_local=None,
        timezone=None,
        capture_datetime_utc=None,
        capture_timezone_source="unknown",
        capture_datetime_source="unknown",
        gps_datetime_utc=None,
        gps_timezone=None,
    )

    for field, value in _non_activity_image_payload(rotated).items():
        setattr(image, field, value)

    await session.commit()
    await session.refresh(image)
    return TripImageRead.model_validate(image)

File: apps/web/src/lib/trips.ts
import { CommentRead } from "@/lib/comments";

export type TripContent = Record<string, unknown> | null;

export type TripImageRead = {
  id: number;
  trip_id: number;
  position: number;
  image_url: string;
  thumbnail_url: string;
  tiny_thumbnail_url: string | null;
  width: number;
  height: number;
  thumbnail_width: number;
  thumbnail_height: number;
  tiny_thumbnail_width: number | null;
  tiny_thumbnail_height: number | null;
  content_type: string;
  original_filename: string | null;
  gps_latitude: number | null;
  gps_longitude: number | null;
  created_at: string;
};

export type TripRead = {
  id: number;
  name: string;
  content: TripContent;
  start_date: string | null;
  end_date: string | null;
  timezone: string | null;
  country_codes: string[];
  planned_distance_m: number | null;
  planned_path_polyline: string | null;
  show_planned_path: boolean;
  map_card_image_url: string | null;
  map_card_width: number | null;
  map_card_height: number | null;
  map_card_content_type: string | null;
  comments: CommentRead[];
  latitude: number | null;
  longitude: number | null;
  zoom: number | null;
  metrics_config: Record<string, unknown>;
  images: TripImageRead[];
  created_at: string;
};

export type TripWrite = {
  name: string;
  content: TripContent;
  start_date: string | null;
  end_date: string | null;
  timezone: string | null;
  country_codes: string[];
  planned_distance_m: number | null;
  planned_path_polyline: string | null;
  show_planned_path: boolean;
  latitude: number | null;
  longitude: number | null;
  zoom: number | null;
  metrics_config: Record<string, unknown>;
};

function getApiBaseUrl() {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  if (!baseUrl) {
    throw new Error("Missing NEXT_PUBLIC_API_BASE_URL");
  }
  return baseUrl;
}

async function request<T>(path: string, token: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${getApiBaseUrl()}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      ...init?.headers,
    },
    cache: "no-store",
  });

  if (res.status === 401 || res.status === 403) {
    throw new Error("AUTH_REQUIRED");
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || `Request failed: ${res.status}`);
  }

  if (res.status === 204) {
    return undefined as T;
  }

  return (await res.json()) as T;
}

export function listTrips(token: string) {
  return request<TripRead[]>("/api/v1/trips", token);
}

export function createTrip(token: string, payload: TripWrite) {
  return request<TripRead>("/api/v1/trips", token, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateTrip(token: string, tripId: number, payload: TripWrite) {
  return request<TripRead>(`/api/v1/trips/${tripId}`, token, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteTrip(token: string, tripId: number) {
  return request<void>(`/api/v1/trips/${tripId}`, token, {
    method: "DELETE",
  });
}

export function uploadTripGpx(
  token: string,
  tripId: number,
  payload: {
    file: File;
    compress: boolean;
  },
) {
  const body = new FormData();
  body.append("file", payload.file);
  body.append("compress", String(payload.compress));

  return requestMultipart<TripRead>(`/api/v1/trips/${tripId}/gpx`, token, body, {
    method: "POST",
  });
}

async function requestMultipart<T>(
  path: string,
  token: string,
  body: FormData,
  init?: RequestInit,
): Promise<T> {
  const res = await fetch(`${getApiBaseUrl()}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      ...init?.headers,
    },
    body,
    cache: "no-store",
  });

  if (res.status === 401 || res.status === 403) {
    throw new Error("AUTH_REQUIRED");
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || `Request failed: ${res.status}`);
  }

  if (res.status === 204) {
    return undefined as T;
  }

  return (await res.json()) as T;
}

export function uploadTripImages(
  token: string,
  tripId: number,
  payload: {
    files: File[];
    resizeMode: "keep" | "resize";
    resizeWidth: number | null;
    resizeHeight: number | null;
  },
) {
  const body = new FormData();
  for (const file of payload.files) {
    body.append("files", file);
  }
  body.append("resize_mode", payload.resizeMode);
  if (payload.resizeMode === "resize") {
    body.append("resize_width", String(payload.resizeWidth ?? ""));
    body.append("resize_height", String(payload.resizeHeight ?? ""));
  }

  return requestMultipart<TripImageRead[]>(`/api/v1/trips/${tripId}/images`, token, body, {
    method: "POST",
  });
}

export function deleteTripImage(token: string, tripId: number, imageId: number) {
  return request<void>(`/api/v1/trips/${tripId}/images/${imageId}`, token, {
    method: "DELETE",
  });
}

export function reorderTripImages(token: string, tripId: number, orderedImageIds: number[]) {
  return request<TripImageRead[]>(`/api/v1/trips/${tripId}/images/order`, token, {
    method: "PUT",
    body: JSON.stringify({ ordered_image_ids: orderedImageIds }),
  });
}

export function rotateTripImage(token: string, tripId: number, imageId: number) {
  return request<TripImageRead>(`/api/v1/trips/${tripId}/images/${imageId}/rotate`, token, {
    method: "PATCH",
  });
}

export function uploadTripMapCardImage(
  token: string,
  tripId: number,
  payload: {
    file: File;
    resizeMode: "keep" | "resize";
    resizeWidth: number | null;
    resizeHeight: number | null;
  },
) {
  const body = new FormData();
  body.append("file", payload.file);
  body.append("resize_mode", payload.resizeMode);
  if (payload.resizeMode === "resize") {
    body.append("resize_width", String(payload.resizeWidth ?? ""));
    body.append("resize_height", String(payload.resizeHeight ?? ""));
  }

  return requestMultipart<TripRead>(`/api/v1/trips/${tripId}/map-card-image`, token, body, {
    method: "POST",
  });
}

export function deleteTripMapCardImage(token: string, tripId: number) {
  return request<TripRead>(`/api/v1/trips/${tripId}/map-card-image`, token, {
    method: "DELETE",
  });
}

export function deleteTripComment(token: string, tripId: number, commentId: number) {
  return request<CommentRead>(`/api/v1/trips/${tripId}/comments/${commentId}`, token, {
    method: "DELETE",
  });
}

export function rotateTripMapCardImage(token: string, tripId: number) {
  return request<TripRead>(`/api/v1/trips/${tripId}/map-card-image/rotate`, token, {
    method: "PATCH",
  });
}
